File: sdk/core/src/events.ts
import { PublicKey } from "@solana/web3.js";

export interface TokenInitializedEvent {
  mint: PublicKey;
  authority: PublicKey;
  name: string;
  symbol: string;
  decimals: number;
  preset: number;
  timestamp: bigint;
}

export interface TokensMintedEvent {
  mint: PublicKey;
  recipient: PublicKey;
  amount: bigint;
  minter: PublicKey;
  totalMinted: bigint;
  timestamp: bigint;
}

export interface TokensBurnedEvent {
  mint: PublicKey;
  from: PublicKey;
  amount: bigint;
  burner: PublicKey;
  totalBurned: bigint;
  timestamp: bigint;
}

export interface AccountFrozenEvent {
  mint: PublicKey;
  account: PublicKey;
  frozenBy: PublicKey;
  timestamp: bigint;
}

export interface AccountThawedEvent {
  mint: PublicKey;
  account: PublicKey;
  thawedBy: PublicKey;
  timestamp: bigint;
}

export interface AddressBlacklistedEvent {
  config: PublicKey;
  address: PublicKey;
  reason: string;
  blacklistedBy: PublicKey;
  timestamp: bigint;
}

export interface TokensSeizedEvent {
  config: PublicKey;
  from: PublicKey;
  to: PublicKey;
  amount: bigint;
  seizedBy: PublicKey;
  timestamp: bigint;
}

export type StablecoinEvent =
  | { type: "TokenInitialized"; data: TokenInitializedEvent }
  | { type: "TokensMinted"; data: TokensMintedEvent }
  | { type: "TokensBurned"; data: TokensBurnedEvent }
  | { type: "AccountFrozen"; data: AccountFrozenEvent }
  | { type: "AccountThawed"; data: AccountThawedEvent }
  | { type: "AddressBlacklisted"; data: AddressBlacklistedEvent }
  | { type: "TokensSeized"; data: TokensSeizedEvent };

export type EventCallback<T = unknown> = (event: T) => void;

export class EventManager {
  private listeners: Map<string, EventCallback[]> = new Map();

  on<T>(eventType: string, callback: EventCallback<T>): () => void {
    const existing = this.listeners.get(eventType) || [];
    existing.push(callback as EventCallback);
    this.listeners.set(eventType, existing);

    return () => {
      const cbs = this.listeners.get(eventType) || [];
      const idx = cbs.indexOf(callback as EventCallback);
      if (idx >= 0) cbs.splice(idx, 1);
    };
  }

  emit<T>(eventType: string, data: T): void {
    const cbs = this.listeners.get(eventType) || [];
    for (const cb of cbs) {
      try {
        cb(data);
      } catch {
        // swallow listener errors
      }
    }
  }

  removeAll(): void {
    this.listeners.clear();
  }
}

File: programs/sss-token/src/events.rs
use anchor_lang::prelude::*;

#[event]
pub struct TokenInitialized {
    pub mint: Pubkey,
    pub authority: Pubkey,
    pub name: String,
    pub symbol: String,
    pub decimals: u8,
    pub preset: u8,
    pub timestamp: i64,
}

#[event]
pub struct TokensMinted {
    pub mint: Pubkey,
    pub recipient: Pubkey,
    pub amount: u64,
    pub minter: Pubkey,
    pub total_minted: u64,
    pub timestamp: i64,
}

#[event]
pub struct TokensBurned {
    pub mint: Pubkey,
    pub from: Pubkey,
    pub amount: u64,
    pub burner: Pubkey,
    pub total_burned: u64,
    pub timestamp: i64,
}

#[event]
pub struct AccountFrozen {
    pub mint: Pubkey,
    pub account: Pubkey,
    pub frozen_by: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct AccountThawed {
    pub mint: Pubkey,
    pub account: Pubkey,
    pub thawed_by: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct TokenPaused {
    pub mint: Pubkey,
    pub paused_by: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct TokenUnpaused {
    pub mint: Pubkey,
    pub unpaused_by: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct RoleGranted {
    pub config: Pubkey,
    pub user: Pubkey,
    pub role: u8,
    pub granted_by: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct RoleRevoked {
    pub config: Pubkey,
    pub user: Pubkey,
    pub role: u8,
    pub revoked_by: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct MinterQuotaUpdated {
    pub config: Pubkey,
    pub minter: Pubkey,
    pub new_quota: u64,
    pub updated_by: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct AuthorityTransferred {
    pub config: Pubkey,
    pub old_authority: Pubkey,
    pub new_authority: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct AddressBlacklisted {
    pub config: Pubkey,
    pub address: Pubkey,
    pub reason: String,
    pub blacklisted_by: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct AddressUnblacklisted {
    pub config: Pubkey,
    pub address: Pubkey,
    pub removed_by: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct TokensSeized {
    pub config: Pubkey,
    pub from: Pubkey,
    pub to: Pubkey,
    pub amount: u64,
    pub seized_by: Pubkey,
    pub timestamp: i64,
}
